redbot/core/utils/__init__.py

```diff
@@ -255,3 +256,54 @@ def bounded_gather(
     tasks = (_sem_wrapper(semaphore, task) for task in coros_or_futures)
 
     return asyncio.gather(*tasks, return_exceptions=return_exceptions)
+
+
+class AsyncGen(AsyncIterable[_T]):
+    """Yield entry every `delay` seconds per `steps`."""
+
+    def __init__(
+        self, contents: Union[Iterator[_T], Iterable[_T]], delay: float = 0.0, steps: int = 1
+    ) -> None:
+        self.delay = delay
+        self.content = iter(contents)
+        self.i = 0
+        self.steps = steps
+
+    def __aiter__(self):
+        return self
+
+    async def __anext__(self) -> Awaitable[_T]:
+        try:
+            i = next(self.content)
+        except StopIteration:
+            raise StopAsyncIteration
+        self.i += 1
+        if self.i % self.steps == 0:
+            await asyncio.sleep(self.delay)
+        return i
+
+    def __await__(self):
+        return self.flatten()
+
+    async def flatten(self) -> List[_T]:
+        return [item async for item in self]
+
+    async def filter(
+        self, function: Union[AsyncIterable[_T], Iterable[_T]], *args, **kwargs
+    ) -> AsyncIterator[_T]:
+        function = partial(function, *args, **kwargs)
+        async for i in async_filter(self, function):
+            yield i
+
+    async def enumerate(self, start=0) -> AsyncIterator[Tuple[int, _T]]:
+        async for item in self:
+            yield start, item
+            start += 1
```

Comment:
- I think using `async_enumerate` would be better here just so that all the logic is in one place.
- `start` is missing type hint
```suggestion
    def enumerate(self, start: int = 0) -> AsyncIterator[Tuple[int, _T]]:
        return async_enumerate(self, start)
```